### PR TITLE
feat(config): call setup() from per-plugin opts

### PR DESCRIPTION
**Pull request by yukimemi**
## What

Adds **`opts`** and **`main`** to `[[plugins]]`, so a plugin whose README says
"call `require('x').setup({})` yourself" no longer needs a hand-written,
setup-only `after.lua`.

```toml
[[plugins]]
url      = "https://github.com/yukimemi/autocursor.nvim"
on_event = ["BufReadPre", "BufNewFile"]
opts     = {}                                   # → require("autocursor").setup({})
# opts = { throttle = 500, cursorline = { enable = true } }
# main = "autocursor"                            # resolution override, normally omitted
```

The **presence** of `opts` is the switch — lazy.nvim's convention, verified
against its source (`lazy/core/loader.lua`: `if plugin.config or plugin.opts then
M.config(plugin) end`, i.e. lazy.nvim does *not* call setup without `opts`
either). An absent field keeps today's behavior of never calling setup.

## Why this shape

`options.auto_setup` (a global "call setup for everything" flag) was considered
and rejected: plugins without a `setup()`, plugins that call setup from their own
`plugin/*.lua`, and setups with required arguments would all break, and behavior
would hinge on per-plugin directory heuristics — the opposite of rvpm's
"loader.lua is explicit and deterministic" promise. Per-plugin opt-in data keeps
the decision in `config.toml`, visible in diffs.

## Implementation notes

- **Resolution is ahead-of-time.** `plugin_scan::resolve_main_module` implements
  lazy.nvim's `Loader.get_main` rules (explicit `main` → `mini.*` special case →
  `normname` match against `lua/` top-level modules → sole module → give up) at
  *generate* time. loader.lua receives a finished literal
  (`rvpm_setup("<name>", "<module>", { ... })`), so startup pays nothing for the
  lookup lazy.nvim performs at runtime, and an unresolvable module is a
  generate-time warning (asking for `main = "..."`) instead of a runtime error.
- **Call position**: after `after/plugin/**`, immediately before the plugin's
  `after.lua` — phase 6 for eager plugins, the identical slot inside `load_lazy`
  for lazy ones (new 9th positional argument). Order is always
  **setup → after.lua**, mirroring lazy.nvim's `opts` → `config`.
- **Lazy stays lazy**: the setup is wrapped in a closure
  (`local _rvpm_st_<name> = function() rvpm_setup(...) end`), so the `opts` table
  literal is never constructed unless the trigger fires. Lazy deps get the same
  treatment.
- **Resilience**: `rvpm_setup` `pcall`s the call and reports via
  `vim.notify(..., vim.log.levels.ERROR)`. A throwing `setup()` must not take
  down the remaining eager plugins or the phase 7 trigger registrations.
- **`opts` strings are not path-normalized**: `lua_quote` rewrites `\` → `/`
  for paths, so `opts` values go through a separate faithful quoter. Non-finite
  TOML floats (`inf` / `nan`) map to `math.huge` / `0/0`, and non-identifier or
  keyword keys are emitted as `["end"] = ...`.
- **Frontends**: `rvpm add` suggests `opts = {}` when the resolved module
  declares `setup` (gated by `options.auto_lazy`, but *independent* of the
  trigger half — `--no-lazy` still gets the suggestion). `src/ai/schema_prompt.md`
  now prefers `opts` over a setup-only `after.lua` and, on `rvpm tune`, folds an
  existing setup-only `after.lua` into `opts`. `rvpm set` deliberately leaves
  `opts` alone.
- **Guardrail**: generate warns when `opts` is set *and* `before.lua` /
  `after.lua` also calls `.setup(` (double setup).
- **Known limit**: TOML holds data only. Callbacks stay in `after.lua`, the same
  escape hatch lazy.nvim users reach for with `config = function`.

## Verification

`cargo make check` green (fmt / clippy / 923 lib tests + doc-tests / lock-check).

End-to-end against the real `yukimemi/autocursor.nvim` clone, driving actual
Neovim 0.12.4 with the generated loader.lua:

| scenario | result |
| --- | --- |
| lazy (`on_event`) + nested/array opts | not loaded at startup, loaded after `BufReadPre`; `require("autocursor.config").options` = `throttle = 111`, `notify = false`, `ignore_filetypes = { "qf", "help" }`, `cursorline.enable = true`; 2 `CursorHold` autocmds installed (setup really ran) |
| eager + keyword key | loaded at startup, `throttle = 222`, `["end"]` arrived as a real Lua key |
| `main` pointing at a missing module | `rvpm: setup() failed for autocursor.nvim: module ... not found` via `vim.notify`, and loader.lua **kept going** — the plugin's `after.lua` still ran (`throttle = 333`) |
| `opts` + `after.lua` both calling setup | generate-time warning naming the hook path |

## Follow-up

`yukimemi/autocursor.nvim`'s README can drop its "rvpm doesn't auto-run
`setup()`" note in favor of `opts = {}` once this ships.

## Summary by CodeRabbit

* **New Features**
  * Added per-plugin `opts` configuration for automatic Lua module setup.
  * Added optional main-module overrides and automatic module detection.
  * Setup runs in the correct order for eager and lazy plugins, including dependencies.
  * `rvpm add` can suggest setup configuration and generate an empty `opts` table.
  * Added warnings for unresolved modules and duplicate setup calls.
* **Documentation**
  * Expanded plugin configuration, CLI, migration, setup ordering, and hook guidance.
* **Tests**
  * Added coverage for option serialization, module detection, setup ordering, lazy loading, and dependency handling.

### CODE REVIEW

File: src/loader.rs [hunk 1]
```diff
@@ -185,6 +202,106 @@ pub(crate) fn lua_quote(s: &str) -> String {
...
+    for c in s.chars() {
+        match c {
+            '\\' => out.push_str("\\\\"),
+            '"' => out.push_str("\\\""),
+            '\n' => out.push_str("\\n"),
+            '\r' => out.push_str("\\r"),
+            '\t' => out.push_str("\\t"),
+            // その他の制御文字は decimal escape (`\ddd`) で表現する。
+            c if (c as u32) < 0x20 || c as u32 == 0x7f => {
+                out.push_str(&format!("\\{}", c as u32));
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/loader.rs]: @claude Fixed in `2c84fc1` — `format!("\\{:03}", c as u32)`, so control characters now emit `\001` / `\127` with a fixed 3-digit boundary.

Confirmed the bug and the fix in the real runtime (LuaJIT via `nvim --headless`):

```
"\0012" -> #s = 2, bytes 1, 50   (SOH + '2', intended)
"\12"   -> #s = 1, byte 12       (form feed, the old corrupted output)
```

Regression test added next to the backslash test: `loader::tests::toml_to_lua_zero_pads_control_character_escapes` covers `"\u00012"` (control char immediately followed by a literal digit), a lone `\u0001`, DEL (`\u007F`), and asserts `\t` / `\n` still use the readable named escapes.

File: src/plugin_build.rs [hunk 2]
```diff
@@ -418,9 +422,51 @@ pub(crate) fn build_plugin_scripts(
...
 }
 
+/// `opts` から `SetupSpec` (require する module + Lua literal 化した opts) を決める。
+///
+/// - `opts` 未指定 → `None` (rvpm は setup を呼ばない = 従来どおり hook 任せ)
+/// - main module が解決できない → warn して `None`。 `main = "..."` の明示を促す。
+///   resilience: ここで落とさず、その plugin の setup だけを諦める。
+/// - hook (`before.lua` / `after.lua`) が既に `.setup(` を呼んでいそうなら warn。
+///   二重 setup は plugin 側の内部状態を壊しうるので黙って重ねない。
+fn resolve_setup(
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/plugin_build.rs:436]: @claude Agreed — added six tests in `src/plugin_build.rs` (`2c84fc1`), and extracted the double-setup predicate into `hook_calls_setup()` so the matching itself is directly testable rather than only reachable through the `eprintln!` branch:

- `resolve_setup_returns_none_without_opts`
- `resolve_setup_skips_when_module_is_ambiguous` — two candidates under `lua/`, no name match; pins "never silently set up a different module"
- `resolve_setup_honors_explicit_main_over_scanning`
- `resolve_setup_resolves_module_from_lua_tree` — `autocursor.nvim` → `autocursor`
- `resolve_setup_still_emits_when_a_hook_also_calls_setup` — pins that the warning does *not* silently disable `opts` (the user decides which one to drop)
- `hook_calls_setup_detects_only_real_calls` — positive, negative, and unreadable-path cases

`cargo test --lib` is 930 passing.

File: docs/cli.md [hunk 3]
```diff
@@ -11,12 +11,12 @@ behavior and per-command edge cases.
...
...uto-lazy` / `--no-lazy` overrides) chooses an interactive prompt / unconditional accept / skip (`AutoLazyPolicy::Ask/Always/Never`). On accept, `suggest_cmd_triggers_smart` LCP-clusters command groups (regex-izes them as `/^Prefix/` when there is a 3+ character common prefix); keymaps are enumerated; the corresponding `[[plugins]]` entry in `config.toml` gets `on_cmd` / `on_map` patched in place. The same scan also drives the **`opts = {}` suggestion**: when the plugin's main module resolves (`plugin_scan::resolve_main_module`, generate-time rules) and that module statically declares `setup` (`has_setup_function`), rvpm offers to write `opts = {}` into the entry so it calls `require("<module>").setup({})` for you. It shares the one prompt with the trigger suggestion and obeys the same `options.auto_lazy` policy (`ask` / `always` / `never`) plus the `--auto-lazy` / `--no-lazy` overrides; `never` / `--no-lazy` skip the scan (and therefore the `opts` suggestion) entirely. An explicitly eager entry (`lazy = false`) drops only the trigger half — the `opts` suggestion still shows, because an eager plugin needs its `setup()` just as much. Entries that already have `opts` are left alone. |
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply docs/cli.md:14]: @coderabbitai Declining this one — the current doc text matches the code.

`--no-lazy` is not an "eager" flag; it is a policy override that maps to `AutoLazyPolicy::Never` (`src/cli.rs:515-519`):

```rust
let policy_override = if auto_lazy {
    Some(crate::config::AutoLazyPolicy::Always)
} else if no_lazy {
    Some(crate::config::AutoLazyPolicy::Never)
} else {
    None
};
```

and `run_add` gates the whole scan on that policy (`src/commands/add.rs`: `if !matches!(policy, AutoLazyPolicy::Never)`), so `--no-lazy` does skip the scan — including the `opts` suggestion. Its own help text says exactly that: "Skip the auto-scan entirely for this invocation. Overrides `options.auto_lazy` for this call (== \"never\")".

The case you are describing — eager, but still offered `opts = {}` — is an entry with explicit `lazy = false` (`plugin.lazy_raw == Some(false)` → `triggers_allowed = false`), which is what the test at `add.rs` `build_add_suggestion_keeps_opts_when_triggers_are_disallowed` covers. docs/cli.md already documents both halves separately ("`never` / `--no-lazy` skip the scan entirely; explicit `lazy = false` drops only the trigger half"), so no change needed.

File: src/loader.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/loader.rs]: @coderabbitai Fixed in `2c84fc1`, exactly as proposed (`\{:03}`).

Your probe matches what the real runtime does — verified with LuaJIT through `nvim --headless`:

```
"\0012" -> #s = 2, bytes 1, 50   (SOH + '2')
"\12"   -> #s = 1, byte 12       (form feed — the old output)
```

Regression test added first as requested: `loader::tests::toml_to_lua_zero_pads_control_character_escapes` (control char + digit, lone control char, DEL, plus `\t` / `\n` keeping their named escapes).

[the remark this answers is not in the document]
Comment (yukimemi) [reply src/loader.rs]: @claude Your checklist mentioned a third item — "ColorSchemePre missing setup arg" — and you were right to look: that callsite **was** missed. Fixed in `df4ba9b`.

`src/loader.rs` phase 8 built its own `load_lazy(...)` string with only 8 arguments, because the handler lives outside the per-plugin `do ... end` scope and therefore cannot reference `_rvpm_st_<name>`. A lazy colorscheme carrying `opts` would have loaded via `:colorscheme <name>` with `setup = nil` — i.e. every trigger path called setup except that one.

The setup closure is now inlined there, the same way `plugin_files` / `denops_plugins` already are:

```lua
["tokyonight"] = function() load_lazy("tokyonight.nvim", "...", { ... }, {}, {}, nil, nil, {}, function() rvpm_setup("tokyonight.nvim", "tokyonight", { style = "night" }) end) end
```

Two tests pin it: `lazy_colorscheme_handler_passes_setup_closure` and `lazy_colorscheme_handler_passes_nil_setup_without_opts` (the latter guards the `nil` case so the argument position cannot silently shift).

`cargo make check` green: 932 lib tests + doc-tests, clippy, fmt, lock-check.
